postgis/lwgeom_ogc.c

```diff
@@ -684,27 +652,17 @@ Datum LWGEOM_z_point(PG_FUNCTION_ARGS)
 PG_FUNCTION_INFO_V1(LWGEOM_m_point);
 Datum LWGEOM_m_point(PG_FUNCTION_ARGS)
 {
-	GSERIALIZED *geom;
-	LWPOINT *point = NULL;
-	LWGEOM *lwgeom;
-	POINT3DM p;
-
-	geom = PG_GETARG_GSERIALIZED_P(0);
+	GSERIALIZED *geom = PG_GETARG_GSERIALIZED_P(0);
+	POINT4D pt;
 
-	if ( gserialized_get_type(geom) != POINTTYPE )
+	if (gserialized_get_type(geom) != POINTTYPE)
 		lwpgerror("Argument to ST_M() must be a point");
 
-	lwgeom = lwgeom_from_gserialized(geom);
-	point = lwgeom_as_lwpoint(lwgeom);
-
-	if (lwgeom_is_empty(lwgeom) || !lwgeom_has_m(lwgeom))
+	if (!gserialized_has_m(geom) || (gserialized_peek_first_point(geom, &pt) == LW_FAILURE))
```

Comment:
is compiler clever enough to inline it and skip unpacking of everything except M?

Reply:
I don't think so, it is going to bring the 4 coordinates if available. Initially I was going to write 4 different functions instead, but the minimal performance gain wasn't worth duplicating the code (4 coordinates x 2 serialization methods).

Reply:
do we get .bc files for liblwgeom so that postgres's JIT can inline them in runtime?

Reply:
> do we get .bc files for liblwgeom so that postgres's JIT can inline them in runtime?

AFAIK, no since `liblwgeom` isn't compiled using postgres machinery.